rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp

```diff
@@ -236,6 +236,7 @@ __check_parameters(
     } else {
       descriptor = parameter_infos[name].descriptor;
     }
+    descriptor.name = name;
```

Comment:
this will change the behavior to overwrite the `descriptor.name` from user application.

before, can be `what():  parameter 'foo' could not be set: Parameter {bar} doesn't comply with integer range.`, if user put the descriptor name into `bar`. but after this PR, it will be `what():  parameter 'foo' could not be set: Parameter {foo} doesn't comply with integer range.`. so it cannot print the descriptor name.

probably we would want to store actual parameter name in here if `descriptor.name` is empty? or maybe overwriting would be okay? i would like to hear from other people on this.

Reply:
Good point! I forgot that this descriptor is provided by the user.

> or maybe overwriting would be okay?

Maybe. It's unclear to me what the expectation with the parameter descriptors are. Do they need to be fully populated by the user? What if the user provides a different parameter name? I don't see any code that checks if there is a discrepancy between the parameter name and the name provided in the descriptor, for example.

Reply:
In `declare_parameter(const std::string&)`, it looks like an empty descriptor is provided to satisfy the more complete `declare_parameter` signature. https://github.com/ros2/rclcpp/blob/eac006317692b7eff3aed671eadbff3d72cca098/rclcpp/src/rclcpp/node_interfaces/node_parameters.cpp#L467-L473

That leads me to believe that the `name` argument provided to `declare_parameter` should be the only source of ground truth in this situation, i.e. the `name` argument is the parameter argument, and what is found in the parameter descriptor's name field is irrelevant. I don't have enough context about descriptors to assert this though. Thoughts?

Reply:
i might be wrong...

> Do they need to be fully populated by the user?

AFAIK, i do not think so, this is optional and user can/may put some parts only.

> What if the user provides a different parameter name?

in this case, actual parameter name will be used instead of descriptor name. the same behavior goes to type.

i think that this behavior is intentional, so that user application can use descriptor (for instance of range declaration) on some other parameters as well.

> probably we would want to store actual parameter name in here if descriptor.name is empty?

i am currently inclined to have this one.

- we can keep the user descriptor name if that exists.
- we do not change current behavior.
- we can address the original issue https://github.com/ros2/rclcpp/issues/1906

i am open to hear from other opinion 👂